build-tools/packages/build-tools/src/common/tsCompile.ts

```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type * as tsTypes from "typescript";
+import path from "path";
+import { getTscUtils } from "./tscUtils.js";
+
+// Any paths given by typescript will be normalized to forward slashes.
+// Local paths should be normalized to make any comparisons.
+const directorySeparator = "/";
+const backslashRegExp = /\\/g;
+function normalizeSlashes(path: string): string {
+	return path.includes("\\") ? path.replace(backslashRegExp, directorySeparator) : path;
+}
+
+export function tsCompile({
+	command,
+	cwd,
+	packageJsonTypeOverride,
+}: {
+	command: string;
+	cwd: string;
+	packageJsonTypeOverride?: "commonjs" | "module";
+}): number {
+	// Load the typescript version that is in the cwd scope
+	const tscUtils = getTscUtils(cwd);
+
+	const ts = tscUtils.tsLib;
+	let commandLine = tscUtils.parseCommandLine(command);
+	const diagnostics: tsTypes.Diagnostic[] = [];
+	if (commandLine) {
+		const configFileName = tscUtils.findConfigFile(cwd, commandLine);
+		if (configFileName) {
+			commandLine = ts.getParsedCommandLineOfConfigFile(configFileName, commandLine.options, {
+				...ts.sys,
+				getCurrentDirectory: () => cwd,
+				onUnRecoverableConfigFileDiagnostic: (diagnostic: tsTypes.Diagnostic) => {
+					diagnostics.push(diagnostic);
+				},
+			});
+		} else {
+			throw new Error("Unknown config file in command line");
+		}
+	}
+
+	let code = ts.ExitStatus.DiagnosticsPresent_OutputsSkipped;
+	if (commandLine && diagnostics.length === 0) {
+		const incremental = !!(commandLine.options.incremental || commandLine.options.composite);
+
+		const host = incremental
+			? ts.createIncrementalCompilerHost(commandLine.options)
+			: ts.createCompilerHost(commandLine.options);
```

Comment:
I am guessing you've already checked to see if this conditional is necessary: specifically, that ts.createCompilerHost(..)` does not examine options and return the right kind of compiler on its own.

(ditto for create[Incremental]Program).

Reply:
I had not dug into the internals. I thought I grabbed this from the typescript documentation (https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API), but can no longer find the reference(?!). So, I looked at internals and there is a difference.
The latter `create*Program` code was pre-existing; so, we'll keep that, too.